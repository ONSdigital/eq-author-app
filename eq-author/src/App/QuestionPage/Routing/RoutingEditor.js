import React from "react";
import CustomPropTypes from "custom-prop-types";
import { TransitionGroup } from "react-transition-group";
import styled from "styled-components";
import PropTypes from "prop-types";
import { first, get } from "lodash";

import Loading from "components/Loading";
import RoutingRuleSet from "App/QuestionPage/Routing/RoutingRuleSet";
import RoutingRuleSetMsg from "App/QuestionPage/Routing/RoutingRuleSetMsg";
import Transition from "App/QuestionPage/Routing/Transition";
import { colors } from "constants/theme";

const Title = styled.h2`
  padding: 0.5em 1em;
  color: #666;
  font-size: 1.4em;
  border-bottom: 1px solid ${colors.lightGrey};
  margin: 0;
`;

const Padding = styled.div`
  padding: 2em;
`;

class RoutingEditor extends React.Component {
  static propTypes = {
    currentPage: CustomPropTypes.page.isRequired,
    onAddRoutingRuleSet: PropTypes.func.isRequired,
    onAddRoutingCondition: PropTypes.func.isRequired,
    onDeleteRoutingRule: PropTypes.func.isRequired,
    onUpdateRoutingRule: PropTypes.func.isRequired,
    onUpdateRoutingRuleSet: PropTypes.func.isRequired,
    onDeleteRoutingRuleSet: PropTypes.func.isRequired,
    match: CustomPropTypes.match
  };

  handleElseChange = value => {
    this.props.onUpdateRoutingRuleSet(value);
  };

  handleThenChange = value => {
    this.props.onUpdateRoutingRule(value);
  };

  handleAddCondition = rule => {
    const answer = first(this.props.currentPage.answers);
    this.props.onAddRoutingCondition(rule.id, answer && answer.id);
  };

  handleDeleteRule = rule => {
    const {
      currentPage,
      onDeleteRoutingRule,
      onDeleteRoutingRuleSet
    } = this.props;
    const { routingRuleSet } = currentPage;

    routingRuleSet.routingRules.length > 1
      ? onDeleteRoutingRule(routingRuleSet.id, rule.id)
      : onDeleteRoutingRuleSet(routingRuleSet.id, currentPage.id);
  };

  renderEmptyRuleSet() {
    const { onAddRoutingRuleSet } = this.props;
    return (
      <Transition key="routing-rule-set-empty" exit={false}>
        <RoutingRuleSetMsg
          title="No routing rules exist for this question"
          onAddRuleSet={onAddRoutingRuleSet}
          data-test="routing-rule-set-empty-msg"
        >
          Users completing this question will be taken to the next page.
        </RoutingRuleSetMsg>
      </Transition>
    );
  }

  renderRoutingRuleSet() {
    const { currentPage, match, ...otherProps } = this.props;

    const { routingRuleSet } = currentPage;

    return (
      <Transition key="routing-rule-set" exit={false}>
        <div>
          <RoutingRuleSet
            {...otherProps}
            ruleSet={routingRuleSet}
            onElseChange={this.handleElseChange}
            onAddRoutingCondition={this.handleAddCondition}
            onDeleteRule={this.handleDeleteRule}
            onThenChange={this.handleThenChange}
            match={match}
          />
        </div>
      </Transition>
    );
  }

  render() {
    const { currentPage } = this.props;

    // when new section is added, this component re-renders before
    // the redirect, causing currentPage to be undefined/null
    if (!currentPage) {
      return <Loading height="38rem">Loadingâ€¦</Loading>;
    }

    const { routingRuleSet } = currentPage;

    let content;

    if (routingRuleSet) {
      content = this.renderRoutingRuleSet();
    } else {
      content = this.renderEmptyRuleSet();
    }

    return (
      <div data-test="routing-editor">
        <Title>{get(currentPage, "displayName")}</Title>
        <Padding>
          <TransitionGroup>{content}</TransitionGroup>
        </Padding>
      </div>
    );
  }
}

export default RoutingEditor;
