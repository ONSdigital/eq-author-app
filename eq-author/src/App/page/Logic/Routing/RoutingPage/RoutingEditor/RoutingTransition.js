import React from "react";
import { CSSTransition } from "react-transition-group";
import PropTypes from "prop-types";
import styled from "styled-components";

const timeout = (props) => props.timeout;
const halfTimeout = (props) => props.timeout / 2;

const handleExit = (node) => {
  const { height } = node.getBoundingClientRect();
  node.style.height = `${height}px`;
};

const RoutingTransitionWrapper = styled(CSSTransition).attrs(() => ({
  classNames: "routing",
  onExit: () => handleExit,
}))`
  position: relative;

  &.routing-enter {
    opacity: 0;
    transform: scale(0.9);
    z-index: 200;
  }

  &.routing-enter-active {
    opacity: 1;
    transform: scale(1);
    transition: opacity ${timeout}ms ease-out,
      transform ${timeout}ms cubic-bezier(0.175, 0.885, 0.32, 1.275);
  }

  &.routing-exit {
    opacity: 1;
    transform: scale(1);
  }

  &.routing-exit-active {
    opacity: 0;
    height: 0 !important;
    transform: scale(0.9);
    transition: opacity ${halfTimeout}ms ease-out,
      height ${halfTimeout}ms ease-in ${halfTimeout}ms,
      transform ${halfTimeout}ms ease-in;
  }
`;

const RoutingTransition = ({ children, ...otherProps }) => {
  return (
    <RoutingTransitionWrapper {...otherProps}>
      <div>{children}</div>
    </RoutingTransitionWrapper>
  );
};

RoutingTransition.propTypes = {
  timeout: PropTypes.number,
  children: PropTypes.element,
};

RoutingTransition.defaultProps = {
  timeout: 200,
};

export default RoutingTransition;
