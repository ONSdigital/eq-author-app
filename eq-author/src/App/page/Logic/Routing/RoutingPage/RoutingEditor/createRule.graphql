#import "./RuleEditor/fragment.graphql"

mutation createRoutingRule2($input: CreateRoutingRule2Input!) {
  createRoutingRule2(input: $input) {
    ...RuleEditor
    routing {
      id
      rules {
        id
      }
    }
  }
}