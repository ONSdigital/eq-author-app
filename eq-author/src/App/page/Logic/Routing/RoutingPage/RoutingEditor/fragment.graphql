#import "./RuleEditor/fragment.graphql"

fragment RoutingEditor on Routing2 {
  id
  rules {
    ...RuleEditor
  }
  else {
    id
    logical
    page {
      id
      displayName
    }
    section {
      id
      displayName
    }
    validationErrorInfo {
      id
      errors {
        id
        errorCode
      }
      totalCount
    }
  }
}
