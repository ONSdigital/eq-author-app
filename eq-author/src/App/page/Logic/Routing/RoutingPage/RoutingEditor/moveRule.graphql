mutation moveRoutingRule2($input: MoveRoutingRule2Input!) {
  moveRoutingRule2(input: $input) {
    id
    routing {
      id
      rules {
        id
      }
    }
  }
}
