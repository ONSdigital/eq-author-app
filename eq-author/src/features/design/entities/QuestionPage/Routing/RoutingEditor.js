import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { dropRightWhile, first, last, get, isEmpty } from "lodash";
import { TransitionGroup } from "react-transition-group";

import Transition from "./Transition";
import Loading from "components/Loading";
import RoutingRuleSet from "./RoutingRuleSet";
import RoutingRuleSetMsg from "./RoutingRuleSetMsg";

import { colors } from "constants/theme";
import CustomPropTypes from "custom-prop-types";

const Title = styled.h2`
  padding: 0.5em 1em;
  color: #666;
  font-size: 1.4em;
  border-bottom: 1px solid ${colors.lightGrey};
  margin: 0;
`;

const Padding = styled.div`
  padding: 2em;
`;

const getPagesAvailableForRouting = (sections, sectionId, pageId) => {
  if (isEmpty(sections)) {
    return [];
  }

  const filteredSections = dropRightWhile(sections, s => s.id !== sectionId);
  const currentSection = last(filteredSections);

  filteredSections[filteredSections.length - 1] = {
    ...currentSection,
    pages: dropRightWhile(currentSection.pages, p => p.id !== pageId)
  };

  return filteredSections;
};

class RoutingEditor extends React.Component {
  static propTypes = {
    questionnaire: CustomPropTypes.questionnaire.isRequired,
    currentPage: CustomPropTypes.page.isRequired,
    onAddRoutingRuleSet: PropTypes.func.isRequired,
    onAddRoutingCondition: PropTypes.func.isRequired,
    onDeleteRoutingRule: PropTypes.func.isRequired,
    onUpdateRoutingRule: PropTypes.func.isRequired,
    onUpdateRoutingRuleSet: PropTypes.func.isRequired,
    onDeleteRoutingRuleSet: PropTypes.func.isRequired,
    availableRoutingDestinations: PropTypes.shape({
      logicalDestinations: PropTypes.arrayOf(PropTypes.any),
      questionPages: PropTypes.arrayOf(CustomPropTypes.page),
      sections: PropTypes.arrayOf(CustomPropTypes.section)
    }).isRequired,
    match: CustomPropTypes.match
  };

  handleElseChange = value => {
    this.props.onUpdateRoutingRuleSet(value);
  };

  handleThenChange = value => {
    this.props.onUpdateRoutingRule(value);
  };

  handleAddCondition = rule => {
    const answer = first(this.props.currentPage.answers);
    this.props.onAddRoutingCondition(rule.id, answer && answer.id);
  };

  handleDeleteRule = rule => {
    const {
      currentPage,
      onDeleteRoutingRule,
      onDeleteRoutingRuleSet
    } = this.props;
    const { routingRuleSet } = currentPage;

    routingRuleSet.routingRules.length > 1
      ? onDeleteRoutingRule(routingRuleSet.id, rule.id)
      : onDeleteRoutingRuleSet(routingRuleSet.id, currentPage.id);
  };

  renderEmptyRuleSet() {
    const { onAddRoutingRuleSet } = this.props;
    return (
      <Transition key="routing-rule-set-empty" exit={false}>
        <RoutingRuleSetMsg
          title="No routing rules exist for this question"
          onAddRuleSet={onAddRoutingRuleSet}
          data-test="routing-rule-set-empty-msg"
        >
          Users completing this question will be taken to the next page.
        </RoutingRuleSetMsg>
      </Transition>
    );
  }

  renderRoutingRuleSet() {
    const {
      questionnaire,
      currentPage,
      availableRoutingDestinations,
      match,
      ...otherProps
    } = this.props;

    const pagesAvailableForRouting = getPagesAvailableForRouting(
      questionnaire.sections,
      match.params.sectionId,
      match.params.pageId
    );

    const { routingRuleSet } = currentPage;

    return (
      <Transition key="routing-rule-set" exit={false}>
        <div>
          <RoutingRuleSet
            {...otherProps}
            ruleSet={routingRuleSet}
            destinations={availableRoutingDestinations}
            pagesAvailableForRouting={pagesAvailableForRouting}
            onElseChange={this.handleElseChange}
            onAddRoutingCondition={this.handleAddCondition}
            onDeleteRule={this.handleDeleteRule}
            onThenChange={this.handleThenChange}
            match={match}
          />
        </div>
      </Transition>
    );
  }

  render() {
    const { currentPage } = this.props;

    // when new section is added, this component re-renders before
    // the redirect, causing currentPage to be undefined/null
    if (!currentPage) {
      return <Loading height="38rem">Loadingâ€¦</Loading>;
    }

    const { routingRuleSet } = currentPage;

    let content;

    if (routingRuleSet) {
      content = this.renderRoutingRuleSet();
    } else {
      content = this.renderEmptyRuleSet();
    }

    return (
      <div data-test="routing-editor">
        <Title>{get(currentPage, "displayName")}</Title>
        <Padding>
          <TransitionGroup>{content}</TransitionGroup>
        </Padding>
      </div>
    );
  }
}

export default RoutingEditor;
