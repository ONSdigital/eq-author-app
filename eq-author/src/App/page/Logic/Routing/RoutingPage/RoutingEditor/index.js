import React from "react";
import { propType } from "graphql-anywhere";
import styled from "styled-components";
import PropTypes from "prop-types";
import { flow } from "lodash/fp";

import Button from "components/buttons/Button";
import Reorder from "components/Reorder";

import DestinationSelector from "App/page/Logic/Routing/DestinationSelector";

import transformNestedFragments from "utils/transformNestedFragments";

import fragment from "./fragment.graphql";
import withUpdateRouting from "./withUpdateRouting";
import withCreateRule from "./withCreateRule";
import RuleEditor from "./RuleEditor";
import RoutingTransition from "./RoutingTransition";
import withMoveRule from "./withMoveRule";

export const LABEL_IF = "IF";
export const LABEL_ELSE = "Else";
export const LABEL_ELSE_IF = "ELSE IF";

const AddRuleButton = styled(Button)`
  display: block;
  margin: 2em auto;
  padding: 0.8em 2em;
  border-radius: 2em;
  border-width: 2px;
`;

export class UnwrappedRoutingEditor extends React.Component {
  static fragments = [fragment, ...RuleEditor.fragments];

  static propTypes = {
    routing: propType(transformNestedFragments(fragment, RuleEditor.fragments))
      .isRequired,
    updateRouting: PropTypes.func.isRequired,
    createRule: PropTypes.func.isRequired,
    moveRule: PropTypes.func.isRequired,
  };

  handleAddClick = () => {
    this.props.createRule(this.props.routing.id);
  };

  handleElseChange = (destination) => {
    this.props.updateRouting({
      ...this.props.routing,
      else: destination,
    });
  };

  render() {
    const { routing, moveRule: handleMoveRule } = this.props;

    return (
      <>
        <Reorder
          list={routing.rules}
          onMove={handleMoveRule}
          Transition={RoutingTransition}
        >
          {(props, rule) => (
            <RuleEditor
              routing={props.routing}
              rule={rule}
              key={rule.id}
              ifLabel={
                routing.rules.indexOf(rule) > 0 ? LABEL_ELSE_IF : LABEL_IF
              }
              {...props}
            />
          )}
        </Reorder>
        <AddRuleButton
          variant="secondary"
          small
          onClick={this.handleAddClick}
          data-test="btn-add-rule"
        >
          Add rule
        </AddRuleButton>

        <DestinationSelector
          id="else"
          label={LABEL_ELSE}
          value={routing.else}
          onChange={this.handleElseChange}
          data-test="select-else"
        />
      </>
    );
  }
}

const withMutations = flow(withMoveRule, withUpdateRouting, withCreateRule);

export default withMutations(UnwrappedRoutingEditor);
