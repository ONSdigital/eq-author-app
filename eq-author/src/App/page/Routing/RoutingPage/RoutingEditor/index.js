import React from "react";
import { propType } from "graphql-anywhere";
import styled from "styled-components";
import PropTypes from "prop-types";
import { TransitionGroup } from "react-transition-group";
import { flow } from "lodash/fp";

import Button from "components/buttons/Button";
import IconText from "components/IconText";

import Transition from "App/page/Routing/Transition";
import DestinationSelector from "App/page/Routing/DestinationSelector";
import IconAddRule from "App/page/Routing/icon-add-rule.svg?inline";

import { colors, radius } from "constants/theme";
import transformNestedFragments from "utils/transformNestedFragments";

import fragment from "./fragment.graphql";
import withUpdateRouting from "./withUpdateRouting";
import withCreateRule from "./withCreateRule";
import RuleEditor from "./RuleEditor";

const AddRuleButton = styled(Button)`
  width: 100%;
  margin-bottom: 2em;
  padding: 0.5em;
`;

const Box = styled.div`
  border: 1px solid ${colors.bordersLight};
  border-radius: ${radius};
  opacity: 1;
`;

export class UnwrappedRoutingEditor extends React.Component {
  static fragments = [fragment, ...RuleEditor.fragments];

  static propTypes = {
    routing: propType(transformNestedFragments(fragment, RuleEditor.fragments))
      .isRequired,
    updateRouting: PropTypes.func.isRequired,
    createRule: PropTypes.func.isRequired,
  };

  handleAddClick = () => {
    this.props.createRule(this.props.routing.id);
  };

  handleElseChange = destination => {
    this.props.updateRouting({
      ...this.props.routing,
      else: destination,
    });
  };

  render() {
    const { routing } = this.props;
    return (
      <>
        <TransitionGroup>
          {routing.rules.map((rule, index) => (
            <Transition key={rule.id} exit={false}>
              <RuleEditor
                rule={rule}
                title={index > 0 ? "Or" : null}
                key={rule.id}
              />
            </Transition>
          ))}
        </TransitionGroup>
        <AddRuleButton
          variant="secondary"
          small
          onClick={this.handleAddClick}
          data-test="btn-add-rule"
        >
          <IconText icon={IconAddRule}>Add rule</IconText>
        </AddRuleButton>
        <Box>
          <DestinationSelector
            id="else"
            label="ELSE"
            value={routing.else}
            onChange={this.handleElseChange}
            data-test="select-else"
          />
        </Box>
      </>
    );
  }
}

const withMutations = flow(
  withUpdateRouting,
  withCreateRule
);

export default withMutations(UnwrappedRoutingEditor);
