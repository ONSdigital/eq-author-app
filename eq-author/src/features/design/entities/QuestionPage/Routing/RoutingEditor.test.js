import RoutingEditor from "./RoutingEditor";
import React from "react";
import { shallow } from "enzyme";

const createWrapper = (props, render = shallow) => {
  return render(<RoutingEditor {...props} />);
};

describe("RoutingEditor", () => {
  let wrapper, props;

  const answer = {
    id: "1",
    type: "Number"
  };

  const page = {
    id: "1",
    answers: [answer]
  };

  const section = {
    id: "1",
    pages: [page]
  };

  const questionnaire = {
    id: "1",
    sections: [section]
  };

  const routingRule = {
    id: "1",
    operation: "And",
    conditions: [],
    goto: null
  };

  const routingRuleSet = {
    id: "1",
    routingRules: [],
    questionPage: page,
    else: null
  };

  const availableRoutingDestinations = {
    questionPages: [
      {
        id: "2",
        __typename: "QuestionPage",
        section: {
          id: "1"
        }
      },
      {
        id: "3",
        __typename: "QuestionPage",
        section: {
          id: "1"
        }
      }
    ],
    sections: [
      {
        id: "2",
        __typename: "Section"
      },
      {
        id: "3",
        __typename: "Section"
      }
    ]
  };

  const pageWithRoutingRuleSet = {
    ...page,
    routingRuleSet
  };

  const pageWithRoutingRules = {
    ...pageWithRoutingRuleSet,
    routingRuleSet: {
      id: "2",
      routingRules: [routingRule]
    }
  };

  beforeEach(() => {
    props = {
      questionnaire,
      currentPage: page,
      onAddRoutingRuleSet: jest.fn(),
      onUpdateRoutingRuleSet: jest.fn(),
      onAddRoutingRule: jest.fn(),
      onUpdateRoutingRule: jest.fn(),
      onDeleteRoutingRule: jest.fn(),
      onAddRoutingCondition: jest.fn(),
      onUpdateRoutingCondition: jest.fn(),
      onDeleteRoutingCondition: jest.fn(),
      onToggleConditionOption: jest.fn(),
      onDeleteRoutingRuleSet: jest.fn(),
      availableRoutingDestinations,
      match: {
        params: {
          questionnaireId: questionnaire.id,
          sectionId: section.id,
          pageId: page.id
        }
      }
    };

    wrapper = createWrapper(props);
  });

  describe("when not routing rule set", () => {
    it("should render when no routing rule set", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should allow a rule set to be added", () => {
      wrapper
        .find("[data-test='routing-rule-set-empty-msg']")
        .simulate("addRuleSet");
      expect(props.onAddRoutingRuleSet).toHaveBeenCalled();
    });
  });

  describe("routing rule set", () => {
    beforeEach(() => {
      const withRoutingRuleSet = {
        ...props,
        currentPage: pageWithRoutingRuleSet,
        availableRoutingDestinations
      };

      wrapper = createWrapper(withRoutingRuleSet);
    });

    it("should render routing rule set", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should handle changing else destination", () => {
      const destination = {
        id: "26",
        else: {
          absoluteDestination: {
            destinationType: "QuestionPage",
            destinationId: "19"
          }
        }
      };

      wrapper.find("RoutingRuleSet").simulate("elseChange", destination);
      expect(props.onUpdateRoutingRuleSet).toHaveBeenCalledWith(destination);
    });
  });

  describe("routing rules", () => {
    beforeEach(() => {
      const withRoutingRules = {
        ...props,
        currentPage: pageWithRoutingRules
      };

      wrapper = createWrapper(withRoutingRules);
    });

    it("should render routing rules", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should handle deleting a routing rule", () => {
      const currentPage = {
        ...pageWithRoutingRuleSet,
        routingRuleSet: {
          ...pageWithRoutingRuleSet.routingRuleSet,
          routingRules: [routingRule, routingRule]
        }
      };
      wrapper = createWrapper({
        ...props,
        currentPage
      });

      wrapper.find("RoutingRuleSet").simulate("deleteRule", routingRule);
      expect(props.onDeleteRoutingRule).toHaveBeenCalledWith(
        currentPage.routingRuleSet.id,
        routingRule.id
      );
    });

    it("delete routing rule set when last rule deleted", () => {
      wrapper.find("RoutingRuleSet").simulate("deleteRule", routingRule);
      expect(props.onDeleteRoutingRuleSet).toHaveBeenCalledWith(
        pageWithRoutingRules.routingRuleSet.id,
        pageWithRoutingRules.id
      );
    });

    it("should handle changing the THEN destination", () => {
      const destination = {
        id: "26",
        goto: {
          absoluteDestination: {
            destinationType: "QuestionPage",
            destinationId: "19"
          }
        }
      };

      wrapper.find("RoutingRuleSet").simulate("thenChange", destination);
      expect(props.onUpdateRoutingRule).toHaveBeenCalledWith(destination);
    });

    it("should handle adding a routing condition", () => {
      wrapper
        .find("RoutingRuleSet")
        .simulate("addRoutingCondition", routingRule);

      expect(props.onAddRoutingCondition).toHaveBeenCalledWith(
        routingRule.id,
        page.answers[0].id
      );
    });
  });
});
