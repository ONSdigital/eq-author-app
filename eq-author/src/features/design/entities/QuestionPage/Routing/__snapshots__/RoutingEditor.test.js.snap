// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutingEditor routing rule set should render routing rule set 1`] = `
<div
  data-test="routing-editor"
>
  <RoutingEditor__Title />
  <RoutingEditor__Padding>
    <TransitionGroup
      childFactory={[Function]}
      component="div"
    >
      <Transition__RoutingComponentTransition
        exit={false}
        key="routing-rule-set"
        timeout={200}
      >
        <div>
          <RoutingRuleSet
            destinations={
              Object {
                "questionPages": Array [
                  Object {
                    "__typename": "QuestionPage",
                    "id": "2",
                    "section": Object {
                      "id": "1",
                    },
                  },
                  Object {
                    "__typename": "QuestionPage",
                    "id": "3",
                    "section": Object {
                      "id": "1",
                    },
                  },
                ],
                "sections": Array [
                  Object {
                    "__typename": "Section",
                    "id": "2",
                  },
                  Object {
                    "__typename": "Section",
                    "id": "3",
                  },
                ],
              }
            }
            match={
              Object {
                "params": Object {
                  "pageId": "1",
                  "questionnaireId": "1",
                  "sectionId": "1",
                },
              }
            }
            onAddRoutingCondition={[Function]}
            onAddRoutingRule={[MockFunction]}
            onAddRoutingRuleSet={[MockFunction]}
            onDeleteRoutingCondition={[MockFunction]}
            onDeleteRoutingRule={[MockFunction]}
            onDeleteRoutingRuleSet={[MockFunction]}
            onDeleteRule={[Function]}
            onElseChange={[Function]}
            onThenChange={[Function]}
            onToggleConditionOption={[MockFunction]}
            onUpdateRoutingCondition={[MockFunction]}
            onUpdateRoutingRule={[MockFunction]}
            onUpdateRoutingRuleSet={[MockFunction]}
            pagesAvailableForRouting={
              Array [
                Object {
                  "id": "1",
                  "pages": Array [
                    Object {
                      "answers": Array [
                        Object {
                          "id": "1",
                          "type": "Number",
                        },
                      ],
                      "id": "1",
                    },
                  ],
                },
              ]
            }
            ruleSet={
              Object {
                "else": null,
                "id": "1",
                "questionPage": Object {
                  "answers": Array [
                    Object {
                      "id": "1",
                      "type": "Number",
                    },
                  ],
                  "id": "1",
                },
                "routingRules": Array [],
              }
            }
          />
        </div>
      </Transition__RoutingComponentTransition>
    </TransitionGroup>
  </RoutingEditor__Padding>
</div>
`;

exports[`RoutingEditor routing rules should render routing rules 1`] = `
<div
  data-test="routing-editor"
>
  <RoutingEditor__Title />
  <RoutingEditor__Padding>
    <TransitionGroup
      childFactory={[Function]}
      component="div"
    >
      <Transition__RoutingComponentTransition
        exit={false}
        key="routing-rule-set"
        timeout={200}
      >
        <div>
          <RoutingRuleSet
            destinations={
              Object {
                "questionPages": Array [
                  Object {
                    "__typename": "QuestionPage",
                    "id": "2",
                    "section": Object {
                      "id": "1",
                    },
                  },
                  Object {
                    "__typename": "QuestionPage",
                    "id": "3",
                    "section": Object {
                      "id": "1",
                    },
                  },
                ],
                "sections": Array [
                  Object {
                    "__typename": "Section",
                    "id": "2",
                  },
                  Object {
                    "__typename": "Section",
                    "id": "3",
                  },
                ],
              }
            }
            match={
              Object {
                "params": Object {
                  "pageId": "1",
                  "questionnaireId": "1",
                  "sectionId": "1",
                },
              }
            }
            onAddRoutingCondition={[Function]}
            onAddRoutingRule={[MockFunction]}
            onAddRoutingRuleSet={[MockFunction]}
            onDeleteRoutingCondition={[MockFunction]}
            onDeleteRoutingRule={[MockFunction]}
            onDeleteRoutingRuleSet={[MockFunction]}
            onDeleteRule={[Function]}
            onElseChange={[Function]}
            onThenChange={[Function]}
            onToggleConditionOption={[MockFunction]}
            onUpdateRoutingCondition={[MockFunction]}
            onUpdateRoutingRule={[MockFunction]}
            onUpdateRoutingRuleSet={[MockFunction]}
            pagesAvailableForRouting={
              Array [
                Object {
                  "id": "1",
                  "pages": Array [
                    Object {
                      "answers": Array [
                        Object {
                          "id": "1",
                          "type": "Number",
                        },
                      ],
                      "id": "1",
                    },
                  ],
                },
              ]
            }
            ruleSet={
              Object {
                "id": "2",
                "routingRules": Array [
                  Object {
                    "conditions": Array [],
                    "goto": null,
                    "id": "1",
                    "operation": "And",
                  },
                ],
              }
            }
          />
        </div>
      </Transition__RoutingComponentTransition>
    </TransitionGroup>
  </RoutingEditor__Padding>
</div>
`;

exports[`RoutingEditor when not routing rule set should render when no routing rule set 1`] = `
<div
  data-test="routing-editor"
>
  <RoutingEditor__Title />
  <RoutingEditor__Padding>
    <TransitionGroup
      childFactory={[Function]}
      component="div"
    >
      <Transition__RoutingComponentTransition
        exit={false}
        key="routing-rule-set-empty"
        timeout={200}
      >
        <RoutingRuleSetMsg
          data-test="routing-rule-set-empty-msg"
          onAddRuleSet={[MockFunction]}
          title="No routing rules exist for this question"
        >
          Users completing this question will be taken to the next page.
        </RoutingRuleSetMsg>
      </Transition__RoutingComponentTransition>
    </TransitionGroup>
  </RoutingEditor__Padding>
</div>
`;
